### PR TITLE
Add YoutubeInfoScraper class and tests.  Add method do get Upload-id from a channel-id.

### LINKED ISSUE #11: Interface with Youtube Data API to get channel list.

### PR DESCRIPTION
**Pull request by josefj1519**

### CODE REVIEW

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public final class YoutubeInfoScraperTest {
+
+  private YoutubeInfoScraper scrape;
```
Comment (derrillhd) [inline src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: nit: rename to scraper (noun)

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  private ChannelListResponse mockResponse;
+  private YoutubeResponse apiResponse; 
+    
+  @Before
+  public void setUp() {
+    apiResponse = mock(YoutubeResponse.class);
+    scrape = scrape.setYoutubeAndGetNewInstance(
+        new YouTube.Builder(new NetHttpTransport(),
+        JacksonFactory.getDefaultInstance(),
+        null)
```
Comment (derrillhd) [inline src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: for readability do a comment like:

/* paramName= */ null

so we know what's null

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+public class YoutubeInfoScraper implements YoutubeResponse {
+
+  // TODO: Add seperate file to hold API Key  
+  private String API_KEY = "";
+  private static final String APPLICATION_NAME = "promotube";
+
+  private YouTube youtube;
+
+  private YoutubeInfoScraper(YouTube youtube){
```
Comment (margaretchan) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: nit: whitespace before every open curly brace following style guide https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  * @param ytResponse Pass in this object or any other object that implements YoutubeResponse.
+  * @param channelId Id of a channel.  Can be found in channel's urls.
+  * e.g. https://www.youtube.com/channel/<CHANNEL-ID>
+  * @return string of the channel's upload playlist id.  Null if id is invalid or 
+  * no items were found.
+  */
+  public String scrapeChannelUploadPlaylist(YoutubeResponse ytResponse, String channelId)
+    throws IOException{
+        ChannelListResponse response = ytResponse.getYoutubeChannelListResponse(channelId);
+        if(response==null || response.getItems()==null || response.getItems().size()==0){
```
Comment (margaretchan) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: nit: whitespace between "if" and open parentheses following style guide https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace

Comment (margaretchan) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: nit: `.isEmpty()` is more readable than `.size() == 0`

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.List;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import com.google.sps.youtube.YoutubeResponse;
+import java.util.Properties; 
+import java.io.FileInputStream;
+
+
+public class YoutubeInfoScraper implements YoutubeResponse {
```
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: add java docs

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  // TODO: Add seperate file to hold API Key  
+  private String API_KEY = "";
+  private static final String APPLICATION_NAME = "promotube";
+
+  private YouTube youtube;
```
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: final

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        return channelsInfo.get(0).getContentDetails().getRelatedPlaylists().getUploads();
+   }
+
+  @Override
+  public ChannelListResponse getYoutubeChannelListResponse(String channelId)
+    throws IOException{ 
+        return youtube.channels().list("contentDetails").setId(channelId).execute();
+   }
+
+  public static YoutubeInfoScraper getNewInstance(){
```
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: These two static factory methods aren't really providing anything that the constructors don't already do (unless I'm missing something.)  I would just remove them and call the constructors directly.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      return new YoutubeInfoScraper();
+  }
+
+  public static YoutubeInfoScraper setYoutubeAndGetNewInstance(YouTube youtube){
```
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: see my other comment.

File: src/main/java/com/google/sps/youtube/YoutubeResponse.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+
+package com.google.sps.youtube;
+
+import java.io.IOException;
+import com.google.api.services.youtube.model.ChannelListResponse;
+
+public interface YoutubeResponse{
```
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeResponse.java]: all public classes and interfaces should have java docs

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 6 above)
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: rename to youTubeClient or youTubeService

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  private static final String APPLICATION_NAME = "promotube";
+
+  private YouTube youtube;
+
+  private YoutubeInfoScraper(YouTube youtube){
+      this.youtube = youtube;
+  }
+
+  private YoutubeInfoScraper(){
+    youtube = new YouTube.Builder(new NetHttpTransport(),
```
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: even better is to have your constructors call each other like this:

this(new YouTube.Builder(new NetHttpTransport(),
        JacksonFactory.getDefaultInstance(),
        null)
        .setApplicationName(APPLICATION_NAME)
        .setYouTubeRequestInitializer(new YouTubeRequestInitializer(API_KEY))
        .build())

There's probably a best practice out there to do this.... if I had time to do some google searching.

File: src/main/java/com/google/sps/youtube/YoutubeResponse.java (hunk 9 above)
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeResponse.java]: So I assume you defined this interface to represent your service so you can fake/mock it out?   First I don't think it's necessary because you can just define a mock on the YouTube api client since it's not final (unless you ran into some issue?):

https://developers.google.com/resources/api-libraries/documentation/youtube/v3/java/latest/com/google/api/services/youtube/YouTube.html

So you don't actually need the interface.

But let's say it were final, then you would probably want the interface, but you'd want to name it YouTubeClient or something (the class name implies that it itself is a response, but it has methods that return a response).  It itself isn't a response; it has methods that return a response.

From an OOP perspective, you wouldn't want to have YoutubeInfoScraper implements YouTubeClient (after my rename suggestion) because then you're implying that the Scraper is a YouTubeClient, but it isn't. (See Effective java, prefer composition over inheritance.). It *depends* on a YouTubeClient in which case you'd use DI and inject a fake vs. real YouTubeClient and assign that to a field (which it looks like you're already doing).

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+public final class YoutubeInfoScraperTest {
+
+  private YoutubeInfoScraper scrape;
+  private ChannelListResponse mockResponse;
+  private YoutubeResponse apiResponse; 
+    
+  @Before
+  public void setUp() {
+    apiResponse = mock(YoutubeResponse.class);
+    scrape = scrape.setYoutubeAndGetNewInstance(
```
Comment (derrillhd) [inline src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: don't you want to pass your fake implmentation of the youTube client and not the real one?  You want:


mockYouTubeClient = mock(YouTube.class);
scraper = new Scraper(mockYouTubeClient);

... etc.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        .setYouTubeRequestInitializer(new YouTubeRequestInitializer("API_KEY"))
+        .build());
+  }
+
+  @Test
+  public void scrapeChannelUploadPlaylist_nonExistentChannelId() throws Exception{
+      ChannelListResponse mockResponse = new ChannelListResponse();
+      when(apiResponse.getYoutubeChannelListResponse("CHANNELID_THAT_DOES_NOT_EXIST")).thenReturn(mockResponse.setItems(null));
+      String actual = scrape.scrapeChannelUploadPlaylist(apiResponse, "CHANNELID_THAT_DOES_NOT_EXIST");
+      Assert.assertEquals("null", actual);
```
Comment (derrillhd) [inline src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: nit: you can probably static import Assert since junit is a commonly used utility library. (like java libs and Guava libs)

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 4 above)
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: nulls are discouraged in java in general, so assume that an external library is well-written and doesn't return null unless it explicitly documents that it returns null (via javadocs or @nullable annotations), or it has been observed to return null in some cases.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  * @param channelId Id of a channel.  Can be found in channel's urls.
+  * e.g. https://www.youtube.com/channel/<CHANNEL-ID>
+  * @return string of the channel's upload playlist id.  Null if id is invalid or 
+  * no items were found.
+  */
+  public String scrapeChannelUploadPlaylist(YoutubeResponse ytResponse, String channelId)
+    throws IOException{
+        ChannelListResponse response = ytResponse.getYoutubeChannelListResponse(channelId);
+        if(response==null || response.getItems()==null || response.getItems().size()==0){
+            return "null";
```
Comment (derrillhd) [inline src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Returning a string containing "null" is something that needs to be parsed by the client in the event that it actually is "null", so you really almost never want to do this.

You probably want to make this method @Nullable and actually return null, or even better don't use nulls and return Optional<String> and in this case you'd return Optional.empty() since nulls are discouraged in java.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+ @Test
+  public void scrapeChannelUploadPlaylist_emptyList() throws Exception{
+      ChannelListResponse mockResponse = new ChannelListResponse();
+      when(apiResponse.getYoutubeChannelListResponse("CHANNELID")).thenReturn(mockResponse.setItems(Arrays.asList()));
```
Comment (derrillhd) [inline src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: extract "CHANNELID" into a constant since you're referencing to it multiple times.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+      String actual = scrape.scrapeChannelUploadPlaylist(apiResponse, "CHANNELID");
+      Assert.assertEquals("null", actual);
+  }
+
+  @Test
+  public void scrapeChannelUploadPlaylist_channelExists() throws Exception{
+      mockResponse = new ChannelListResponse();
+      Channel channel = new Channel();
+      channel.setContentDetails(new ChannelContentDetails().setRelatedPlaylists(
+          new RelatedPlaylists().setUploads("UPLOAD_ID")));
```
Comment (derrillhd) [inline src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: introduce more constants for this string literals to stay consistent

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+      mockResponse.setItems(Arrays.asList(channel));
+      when(apiResponse.getYoutubeChannelListResponse("CHANNELID_THAT_EXISTS")).thenReturn(mockResponse);
+      String actual = scrape.scrapeChannelUploadPlaylist(apiResponse, "CHANNELID_THAT_EXISTS");
+      Assert.assertEquals("UPLOAD_ID", actual);
+  }
+
+  @Test (expected = IOException.class)
```
Comment (derrillhd) [inline src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: I would say go see: go/java-testing/exceptions 
:'(

but you can't so I'll just paste it below.

TL;DR is Use assertThrows, and all other ways to verify exceptions are discouraged.

---
Use assertThrows
The strongly preferred solution is to use Assert.assertThrows():

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class AppendOnlyListTest {
  @Test
  public void removeFails() {
    AppendOnlyList list = new AppendOnlyList();
    list.add(0, "a");

    assertThrows(
        UnsupportedOperationException.class,
        () -> list.remove(0));
  }
}
The body of the lambda expression should always be a single statement, not a block.

Testing further details about the thrown exception
In some situations you will decide it is important to check additional details about the thrown exception. You might care about its message, or the attributes of its chained exception, for example.

It is easy to check these details, as assertThrows returns the thrown exception to you, and you can pass it to Truth:

  UnsupportedOperationException thrown =
      assertThrows(
          UnsupportedOperationException.class,
          () -> list.remove(0));
  assertThat(thrown).hasCauseThat().isInstanceOf(SomeOtherException.class);
  assertThat(thrown).hasMessageThat().contains("blah blah blah");
This test will fail if no exception is thrown, if the wrong type is thrown, or if either of the additional assertions on the exception fail.

Supplying additional message text
If you have any useful additional data to provide that will help the engineer debugging a failure, provide a custom message to assertThrows.

  private void assertBad(String input) {
    assertThrows(
        "input was: " + input,
        IllegalArgumentException.class,
        () -> process(input));
  }
Supplying a constant string adds little value, since a failure here always means the same thing: the method just called wasn't expected to complete normally, but it did.

Using the legacy try idiom
There is one case where you might choose not to use assertThrows:

When you aren't able to use Java 8 constructs, particularly lambda expressions. You can still use assertThrows, but you may find it bulky enough to outweigh its advantages.
However, use caution when doing so (read the Warnings below).

Example
import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.fail;

import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class AppendOnlyListTest {
  @Test
  public void removeFails() {
    AppendOnlyList list = new AppendOnlyList();
    list.add(0, "a");

    try {
      list.remove(0);
      fail(); // supply a message IF there is useful data to include
    } catch (UnsupportedOperationException expected) {
    }
  }
}
If you choose to provide a custom message to fail, see the advice above.

Warnings
Never use this idiom to test for AssertionError; it will malfunction and vacuously pass. Note that when testing production code, it should be impossible to cause an AssertionError to happen anyway. But unfortunately, this is also the exception type that some test libraries started using. Even more unfortunately, JUnit 4's regular assertion failures started using it, which is what causes the malfunction described here. Stick to assertThrows if at all possible. (If not, you will need to return from the catch block and move the fail after that.)
Like our advice for assertThrows, you should include only one statement in the try block.
Be careful not to forget the fail(), or your test will vacuously pass. (We try to trigger a warning in Critique in this case, but it doesn't catch everything.)
Use a variable name beginning with expected, as shown, or you will get a lint warning.
Note that all of the above problems can be avoided by using assertThrows.

All other solutions are discouraged
Don't use ExpectedException, @Test(expected =), or ExceptionExpectations. You can find a lengthy explanation of the pitfalls of these approaches in this page's history.

Migrate tests using these legacy utilities onto assertThrows using go/exception-test-migrator.

More advice
If your method should throw FooException, test for FooException
Don't test for a more generic type like RuntimeException or Exception. A test that expects RuntimeException can falsely pass if it contains a bug that triggers a NullPointerException.

When appropriate, perform additional assertions on the exception
to verify that the message contains diagnostic information -- for example, the value that failed its precondition check
to test that the exception happened for one reason and not another -- for example, that an IllegalArgumentException was caused by a -1 parameter and not by a "" parameter
To access an exception's cause consider Throwables.getCauseAs()
In addition to Truth's assertThat(e).hasCauseThat().... assertions, you may want to directly retrieve an exceptions' expected cause. Avoid doing it this way, however:

MyExpectedCause cause = (MyExpectedCause) thrown.getCause();
The problem is that if the cause is not of type MyExpectedCause, the original exception and stack trace will vanish in favor of the new ClassCastException. Instead, prefer Throwables.getCauseAs(), which will always safely preserve the original exception:

MyExpectedCause cause = Throwables.getCauseAs(thrown, MyExpectedCause.class);
Consider FutureSubject when asserting on failed futures
Although you can use assertThrows(ExecutionException.class, someFuture::get) to assert that a future fails, FutureSubject may be easier to use. There is also LabsFutures.getFailureCauseAs() to get the cause of the failed future.

**Review by derrillhd — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 3 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 4 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java (hunk 1 above)
Comment (josefj1519) [reply src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: Done.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java (hunk 2 above)
Comment (josefj1519) [reply src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: Removed this line in the test, added the comment in the YoutubeInfoScraper class to make it more clear.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 5 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 6 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 7 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 8 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

File: src/main/java/com/google/sps/youtube/YoutubeResponse.java (hunk 9 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeResponse.java]: Removed interface.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 6 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 10 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Done.  I was trying to find a way to have default parameters in Java.  I think this is as close as I can get.

File: src/main/java/com/google/sps/youtube/YoutubeResponse.java (hunk 9 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeResponse.java]: Removed the interface.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java (hunk 11 above)
Comment (josefj1519) [reply src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: Done.  Like we had in our discussion during the pod sync.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java (hunk 12 above)
Comment (josefj1519) [reply src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: Done. I assume you mean assertEquals?

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 4 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: https://developers.google.com/resources/api-libraries/documentation/youtube/v3/java/latest/com/google/api/services/youtube/model/ChannelListResponse.html#getItems--
getItems() can return null.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 13 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Changed to optionals.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java (hunk 14 above)
Comment (josefj1519) [reply src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: Done.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java (hunk 15 above)
Comment (josefj1519) [reply src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: Done.

File: src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java (hunk 16 above)
Comment (josefj1519) [reply src/test/java/com/google/sps/youtube/YoutubeInfoScraperTest.java]: Changed to Asserthrows.  Had to update junit version in maven.

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public final class YoutubeInfoScraperTest {
+
+  private YoutubeInfoScraper scraper;
+  private ChannelListResponse mockResponse; 
+  private YouTube mockYouTubeClient;
+  private YouTube.Channels mockChannels;
+  private YouTube.Channels.List mockListChannels;
```
Comment (derrillhd) [inline src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java:54]: only this mock needs to be in a field since it's the only one referenced outside of setUp.  the rest can be local variables to setUp

Comment (derrillhd) [reply src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java:54]: (intellij would have pointed this out to you.)

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+      Optional<String> actual = scraper.scrapeChannelUploadPlaylist(CHANNEL_ID_THAT_DOES_NOT_EXIST);
+      assertEquals(false, actual.isPresent());
+  }
+
+ @Test
+  public void scrapeChannelUploadPlaylist_emptyList() throws Exception{
+      ChannelListResponse mockResponse = new ChannelListResponse();
+      when(mockListChannels.execute()).thenReturn(mockResponse.setItems(Arrays.asList()));
+      Optional<String> actual = scraper.scrapeChannelUploadPlaylist(CHANNEL_ID);
+      assertEquals(false, actual.isPresent());
```
Comment (derrillhd) [inline src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java:86]: optional: consider either using a Matcher: http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertThat(T,%20org.hamcrest.Matcher) or even better adopt https://github.com/google/truth but there's a small learning curve there and you need to add yet another dep into your pom.xml file.

If you adopt the truth library, it would replace all calls to the org.junit.Assert library.  (If your team does decide to adopt, you can do it piecemeal, but make sure each file picks one standard. don't mix truth and non-truth in any single file.)

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 4 above)
Comment (derrillhd) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: execute() doesn't: https://googleapis.dev/java/google-api-client/latest/com/google/api/client/googleapis/services/AbstractGoogleClientRequest.html#execute--
so the first null check is unnecessary.  (unless you observed this behavior in which case document the null happened empirically)

Please document that getItems() returns null when no items match criteria.  you may add the link to the java docs if you like.

File: src/main/java/com/google/step/youtube/YoutubeInfoScraper.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+  * @param ytResponse Pass in this object or any other object that implements YoutubeResponse.
+  * @param channelId Id of a channel.  Can be found in channel's urls.
+  * e.g. https://www.youtube.com/channel/<CHANNEL-ID>
+  * @return an optional string of the channel's upload playlist id. The optional will be empty 
+  * if id is invalid or no items were found.
+  */
+  public Optional<String> scrapeChannelUploadPlaylist(String channelId)
+    throws IOException {
+        ChannelListResponse response = youTubeClient.channels().list("contentDetails").setId(channelId).execute();
+        if (response==null || response.getItems()==null || response.getItems().isEmpty()){
```
Comment (derrillhd) [inline src/main/java/com/google/step/youtube/YoutubeInfoScraper.java]: see my other comment (attadched to the old comment)

File: src/main/java/com/google/step/youtube/YoutubeInfoScraper.java [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+            return Optional.empty();
+        }
+        List<Channel> channelsInfo = response.getItems();
+        return Optional.of(channelsInfo.get(0).getContentDetails().getRelatedPlaylists().getUploads());
```
Comment (derrillhd) [inline src/main/java/com/google/step/youtube/YoutubeInfoScraper.java:57]: why are you only getting the first one?  what do you do with the rest?

**Review by derrillhd — APPROVED** [review]

File: src/main/java/com/google/step/youtube/YoutubeInfoScraper.java (hunk 20 above)
Comment (derrillhd) [reply src/main/java/com/google/step/youtube/YoutubeInfoScraper.java:57]: document why.

File: src/main/java/com/google/step/youtube/YoutubeInfoScraper.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    this(new YouTube.Builder(new NetHttpTransport(),
+        JacksonFactory.getDefaultInstance(),
+        /* httpRequestInitializer= */ null)
+        .setApplicationName(APPLICATION_NAME)
+        .setYouTubeRequestInitializer(new YouTubeRequestInitializer(API_KEY))
+        .build());
+  }
+
+  /**
+  * @param ytResponse Pass in this object or any other object that implements YoutubeResponse.
```
Comment (dallen68) [inline src/main/java/com/google/step/youtube/YoutubeInfoScraper.java]: This is not a param

**Review by dallen68 — APPROVED** [review]

File: src/main/java/com/google/step/youtube/YoutubeInfoScraper.java (hunk 21 above)
Comment (josefj1519) [reply src/main/java/com/google/step/youtube/YoutubeInfoScraper.java]: Done.

File: src/main/java/com/google/step/youtube/YoutubeInfoScraper.java (hunk 20 above)
Comment (josefj1519) [reply src/main/java/com/google/step/youtube/YoutubeInfoScraper.java:57]: We are requesting only one channel-id in the setId() param, so we expect one result on that channel back.

Comment (josefj1519) [reply src/main/java/com/google/step/youtube/YoutubeInfoScraper.java:57]: Added documentation.

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java (hunk 17 above)
Comment (josefj1519) [reply src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java:54]: Done.

File: src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java (hunk 4 above)
Comment (josefj1519) [reply src/main/java/com/google/sps/youtube/YoutubeInfoScraper.java]: Documented and removed the first null check.

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java (hunk 18 above)
Comment (josefj1519) [reply src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java:86]: For me I don't I think I need this for my tests.  I showed it to the other interns and they might use it.

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java [hunk 22]
```diff
@@ -64,34 +72,34 @@ public void setUp() {
...
-      when(apiResponse.getYoutubeChannelListResponse("CHANNELID")).thenReturn(mockResponse.setItems(Arrays.asList()));
-      Optional<String> actual = scraper.scrapeChannelUploadPlaylist(apiResponse, "CHANNELID");
-      Assert.assertEquals(false, actual.isPresent());
+      when(apiResponse.getYoutubeChannelListResponse(CHANNEL_ID)).thenReturn(mockResponse.setItems(Arrays.asList()));
+      Optional<String> actual = scraper.scrapeChannelUploadPlaylist(apiResponse, CHANNEL_ID);
+      assertEquals(false, actual.isPresent());
   }
 
   @Test
   public void scrapeChannelUploadPlaylist_channelExists() throws Exception{
```
Comment (margaretchan) [inline src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java]: Prefer throwing the most specific exception.

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java [hunk 23]
```diff
@@ -64,34 +72,34 @@ public void setUp() {
...
-      Optional<String> actual = scraper.scrapeChannelUploadPlaylist(apiResponse, "CHANNELID_THAT_EXISTS");
-      Assert.assertEquals("UPLOAD_ID", actual.get());
+      when(apiResponse.getYoutubeChannelListResponse(CHANNEL_ID_THAT_EXISTS)).thenReturn(mockResponse);
+      Optional<String> actual = scraper.scrapeChannelUploadPlaylist(apiResponse, CHANNEL_ID_THAT_EXISTS);
+      assertEquals(UPLOAD_ID, actual.get());
   }
 
-  @Test (expected = IOException.class)
+  @Test 
   public void scrapeChannelUploadPlaylist_IOException() throws Exception{
```
Comment (margaretchan) [inline src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java]: nit: add space before curly brace

**Review by margaretchan — APPROVED** [review]

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java (hunk 23 above)
Comment (josefj1519) [reply src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java]: Done.

File: src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java (hunk 22 above)
Comment (josefj1519) [reply src/test/java/com/google/step/youtube/YoutubeInfoScraperTest.java]: Done. Added IOException.
